### PR TITLE
Fix: CA does not work properly while using AWS EC2 IMDSv2

### LINKED ISSUE #3592: CA does not work properly while using AWS EC2 IMDSv2 only in EKS
Recently AWS EKS supports EC2 Instance Metadata Service v2.
 - https://aws.amazon.com/about-aws/whats-new/2020/08/amazon-eks-supports-ec2-instance-metadata-service-v2/

In my testing environment, I create a worker node with IMDSv2 only and it requires to use token-backed sessions to access IMDS. 
 - https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html

However with this condition, CA seems cannot unmarshall it. 

```
I1008 18:57:01.160950       1 aws_util.go:150] fetching http://169.254.169.254/latest/dynamic/instance-identity/document
..........
W1008 18:57:01.760556       1 aws_util.go:166] Error unmarshalling http://169.254.169.254/latest/dynamic/instance-identity/document, skip...
```

Check the CA pod, it keeps OOMed and results in CrashLoopBackOff.

```
# kubectl get pod -n kube-system
NAME                                  READY   STATUS             RESTARTS   AGE
cluster-autoscaler-5b5489859f-2pkdt   0/1     CrashLoopBackOff   6          13m
```

```
# kubectl describe pod cluster-autoscaler-5b5489859f-2pkdt -n kube-system
Name:           cluster-autoscaler-5b5489859f-2pkdt
Namespace:      kube-system
Priority:       0
Node:           ip-172-31-23-13.ap-northeast-1.compute.internal/172.31.23.13
Start Time:     Thu, 08 Oct 2020 19:22:15 +0000
Labels:         app=cluster-autoscaler
                pod-template-hash=5b5489859f
Annotations:    kubernetes.io/psp: eks.privileged
                prometheus.io/port: 8085
                prometheus.io/scrape: true
Status:         Running
IP:             172.31.20.73
IPs:            <none>
Controlled By:  ReplicaSet/cluster-autoscaler-5b5489859f
Containers:
  cluster-autoscaler:
    Container ID:  docker://8cea864df872af960650f9f01061ca52e62855f680306238f75a12cbc798f8a5
    Image:         k8s.gcr.io/autoscaling/cluster-autoscaler:v1.15.7
    Image ID:      docker-pullable://k8s.gcr.io/autoscaling/cluster-autoscaler@sha256:6641a69b4ea5f911ccbb11b75b2675261d90bf169f612c9e960f60036336d664
    Port:          <none>
    Host Port:     <none>
    Command:
      ./cluster-autoscaler
      --v=4
      --stderrthreshold=info
      --cloud-provider=aws
      --skip-nodes-with-local-storage=false
      --expander=least-waste
      --node-group-auto-discovery=asg:tag=k8s.io/cluster-autoscaler/enabled,k8s.io/cluster-autoscaler/LAB-EKS-15
      --balance-similar-node-groups
      --skip-nodes-with-system-pods=false
    State:          Terminated
      Reason:       OOMKilled
      Exit Code:    137
      Started:      Thu, 08 Oct 2020 19:32:27 +0000
      Finished:     Thu, 08 Oct 2020 19:33:06 +0000
    Last State:     Terminated
      Reason:       OOMKilled
      Exit Code:    137
      Started:      Thu, 08 Oct 2020 19:29:07 +0000
      Finished:     Thu, 08 Oct 2020 19:29:46 +0000
```

If uses IMDSv1 back, it works without issue as following:

```
I1008 19:05:20.256839       1 aws_util.go:150] fetching http://169.254.169.254/latest/dynamic/instance-identity/document
I1008 19:05:38.256216       1 aws_cloud_provider.go:380] Successfully load 354 EC2 Instance Types [u-9tb1 m5n.8xlarge z1d.12xlarge m5dn.12xlarge m5.12xlarge c5d.4xlarge c5d.xlarge r6g.2xlarge m4.4xlarge c5.24xlarge r3.8xlarge i3en.24xlarge i3.4xlarge a1.xlarge r5ad.large r5dn.metal x1e u-9tb1.metal m5dn.16xlarge r5n.4xlarge t3.small c5n.2xlarge m5ad.large t3.micro c5d.2xlarge c1.xlarge r5a.24xlarge t3.large r6g.metal r5a.xlarge c6g.xlarge i3en.metal g4dn.xlarge r6g.16xlarge c3.large i2.4xlarge r5d.xlarge t4g.small t3a.xlarge c3.8xlarge m5d.4xlarge r5ad.xlarge h1 c5d.18xlarge u-6tb1.metal p2.8xlarge m6g.2xlarge c5d.metal i3en.2xlarge 
........
I1008 19:05:44.609556       1 auto_scaling_groups.go:354] Regenerating instance to ASG map for ASGs: []
I1008 19:05:44.609579       1 aws_manager.go:266] Refreshed ASG list, next refresh after 2020-10-08 19:06:44.609574794 +0000 UTC m=+102.445561263
I1008 19:05:44.609801       1 main.go:271] Registered cleanup signal handler
I1008 19:05:44.610023       1 node_instances_cache.go:156] Start refreshing cloud provider node instances cache
I1008 19:05:44.610039       1 node_instances_cache.go:168] Refresh cloud provider node instances cache finished, refresh took 1.791µs
I1008 19:05:54.610015       1 static_autoscaler.go:187] Starting main loop
I1008 19:05:54.610119       1 utils.go:622] No pod using affinity / antiaffinity found in cluster, disabling affinity predicate for this loop
I1008 19:05:54.610130       1 filter_out_schedulable.go:63] Filtering out schedulables
I1008 19:05:54.610168       1 filter_out_schedulable.go:80] No schedulable pods
I1008 19:05:54.610188       1 static_autoscaler.go:334] No unschedulable pods
I1008 19:05:54.610203       1 static_autoscaler.go:381] Calculating unneeded nodes
```

I suspect CA does not use token-backed sessions to access IMDS.
 - https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/aws/aws_util.go

### PR DESCRIPTION
**Pull request by darkpssngr**
This fixes #3592.
We use the aws sdk to fetch the metadata. The endpoint explicitly set to `"http://169.254.169.254"` so that tests don't fail.

We tested it by applying these [changes](https://github.com/kubernetes/autoscaler/compare/cluster-autoscaler-1.18.3...shreyas-srinivas:imdsv2-1.18.3?expand=1) in 1.18.3 and it worked fine. tested both scale up and scale down.

This was tested in instances with IMDsV1 enabled and disabled.

### CODE REVIEW

**Comment by darkpssngr** [comment]
accepted CLA

**Comment by darkpssngr** [comment]
@jaypipes, is there any other step I need to follow before it's ready for review?

**Comment by darkpssngr** [comment]
@Jeffwan @feiskyer could you please take a look at this? Not sure if something's pending from my end.

**Comment by darkpssngr** [comment]
@gjtempleton can you please take a look at this?

**Review by gjtempleton — COMMENTED** [review]
Thanks for the PR.

I have one minor nit on expanding the docs, but other than that, this looks good.

File: cluster-autoscaler/cloudprovider/aws/README.md [hunk 1]
```diff
@@ -355,3 +355,6 @@ To refresh static list, please run `go run ec2_instance_types/gen.go` under
   `aws:///us-east-1a/i-01234abcdef`.
 * If you want to use regional STS endpoints (e.g. when using VPC endpoint for
   STS) the env `AWS_STS_REGIONAL_ENDPOINTS=regional` should be set.
+* If you want to run it on instances with IMDSv1 disabled make sure your
+  EC2 launch configuration has the setting `Metadata response hop limit` set to `2`.
+  Otherwise, the `/latest/api/token` call will timeout and result in an error.
```
[the remark this answers is not in the document]
Comment (darkpssngr) [reply cluster-autoscaler/cloudprovider/aws/README.md]: Done

**Review by jaypipes — APPROVED** [review]
Nice work on this @darkpssngr.

**Pull request merged**